crypto/compactcert/builder.go

```diff
@@ -0,0 +1,244 @@
+	participants []Participant
+	parttree     *merklearray.Tree
+
+	// Cached cert, if Build() was called and no subsequent
+	// Add() calls were made.
+	cert *Cert
+}
+
+// MkBuilder constructs an empty builder (with no signatures).  The message
+// to be signed, as well as other security parameters, are specified in
+// param.  The participants that will sign the message are in part and
+// parttree.
+func MkBuilder(param Params, part []Participant, parttree *merklearray.Tree) (*Builder, error) {
+	npart := len(part)
+
+	b := &Builder{
+		Params:       param,
+		sigs:         make([]sigslot, npart),
+		signedWeight: 0,
+		participants: part,
+		parttree:     parttree,
+	}
+
+	return b, nil
+}
+
+// Present checks if the builder already contains a signature at a particular
+// offset.
+func (b *Builder) Present(pos uint64) bool {
+	return b.sigs[pos].part.Weight != 0
+}
+
+// Add a signature to the set of signatures available for building a certificate.
+// verifySig should be set to true in production; setting it to false is useful
+// for benchmarking to avoid the cost of signature checks.
+func (b *Builder) Add(pos uint64, sig crypto.OneTimeSignature, verifySig bool) error {
+	if b.Present(pos) {
+		return fmt.Errorf("position %d already added", pos)
+	}
+
+	// Check participants array
+	if pos >= uint64(len(b.participants)) {
+		return fmt.Errorf("pos %d >= len(participants) %d", pos, len(b.participants))
+	}
+
+	p := b.participants[pos]
+
+	if p.Weight == 0 {
+		return fmt.Errorf("position %d has zero weight", pos)
+	}
+
+	// Check signature
+	ephID := basics.OneTimeIDForRound(b.SigRound, p.KeyDilution)
+	if verifySig && !p.PK.Verify(ephID, b.Msg, sig) {
+		return fmt.Errorf("signature does not verify under ID %v", ephID)
+	}
+
+	// Remember the signature
+	b.sigs[pos].part = p
+	b.sigs[pos].Sig.OneTimeSignature = sig
+	b.signedWeight += p.Weight
+	b.cert = nil
+	return nil
+}
+
+// Ready returns whether the certificate is ready to be built.
+func (b *Builder) Ready() bool {
+	return b.signedWeight >= b.Params.ProvenWeight
+}
+
+// SignedWeight returns the total weight of signatures added so far.
+func (b *Builder) SignedWeight() uint64 {
+	return b.signedWeight
+}
+
+//msgp:ignore sigCommit
+type sigCommit []sigslot
+
+func (sc sigCommit) Length() uint64 {
+	return uint64(len(sc))
+}
+
+func (sc sigCommit) Get(pos uint64) (crypto.Hashable, error) {
+	if pos >= uint64(len(sc)) {
+		return nil, fmt.Errorf("pos %d past end %d", pos, len(sc))
+	}
+
+	return &sc[pos].sigslotCommit, nil
+}
+
+// coinIndex returns the position pos in the sigs array such that the sum
+// of all signature weights before pos is less than or equal to coinWeight,
+// but the sum of all signature weights up to and including pos exceeds
+// coinWeight.
+//
+// coinIndex works by doing a binary search on the sigs array.  The caller
+// should invoke it with lo=0 and hi=len(b.sigs).  The caller should make
+// sure that sigs[*].L is initialized before using coinIndex().
```

Comment:
This comment is a bit confused. Not only should it be initialized, it should also be current (an Add will ruin it). Low-key solution: update the comment. Belt-and-suspenders solution in case someone changes the code later in a weird way: have a boolean in Builder that tracks whether .Ls are valid. It's initialized on false. Becomes true after lines 176-178 (though maybe those lines should be their own function). Becomes false in Add (just like b.cert = nil line 107).

Reply:
Indeed, seems easier to just add the runtime check (should be trivially cheap) rather than write a more elaborate comment.  Thanks!